Mathlib/Analysis/Calculus/FirstDerivativeTest.lean

```diff
@@ -0,0 +1,145 @@
+
+  If there exists a positive number `r > 0` such that for every `x` in `(a − r, a)`
+  we have `f′(x) ≥ 0`, and for every `x` in `(a, a + r)` we have `f′(x) ≤ 0`,
+  then `f` has a local maximum at `a`.
+
+* `localMin_of_deriv_Ioo`: The dual of `first_derivative_max`, for minima.
+
+* `localMax_of_deriv`: 1st derivative test for maxima using filters.
+
+## Tags
+
+derivative test, calculus
+-/
+
+open Set Topology
+
+
+ /-- The First-Derivative Test from calculus, maxima version.
+  Suppose `a < b < c`, `f : ℝ → ℝ` is continuous at `b`,
+  the derivative `f'` is nonnegative on `(a,b)`, and
+  the derivative `f'` is nonpositive on `(b,c)`. Then `f` has a local maximum at `a`. -/
+lemma localMax_of_deriv_Ioo {f : ℝ → ℝ} {a b c : ℝ} (g₀ : a < b) (g₁ : b < c)
+    (h : ContinuousAt f b)
+    (hd₀ : DifferentiableOn ℝ f (Set.Ioo a b))
+    (hd₁ : DifferentiableOn ℝ f (Set.Ioo b c))
+    (h₀ :  ∀ x ∈ Set.Ioo a b, 0 ≤ deriv f x)
+    (h₁ :  ∀ x ∈ Set.Ioo b c, deriv f x ≤ 0) : IsLocalMax f b :=
+  have hIoc : ContinuousOn f (Ioc a b) :=
+    Ioo_union_right g₀ ▸ ContinuousOn.union_continuousAt (isOpen_Ioo (a := a) (b := b))
+        (DifferentiableOn.continuousOn hd₀) (by simp_all)
+  have hIco : ContinuousOn f (Ico b c) :=
+    Ioo_union_left g₁ ▸ ContinuousOn.union_continuousAt (isOpen_Ioo (a := b) (b := c))
+        (DifferentiableOn.continuousOn hd₁) (by simp_all)
+  isLocalMax_of_mono_anti g₀ g₁
+    (monotoneOn_of_deriv_nonneg (convex_Ioc a b) hIoc (by simp_all) (by simp_all))
+    (antitoneOn_of_deriv_nonpos (convex_Ico b c) hIco (by simp_all) (by simp_all))
+
+
+/-- The First-Derivative Test from calculus, minima version. -/
+lemma localMin_of_deriv_Ioo {f : ℝ → ℝ} {a b c : ℝ} (h : ContinuousAt f b)
+    (g₀ : a < b) (g₁ : b < c)
+    (hd₀ : DifferentiableOn ℝ f (Set.Ioo a b)) (hd₁ : DifferentiableOn ℝ f (Set.Ioo b c))
+    (h₀ : ∀ x ∈ Set.Ioo a b, deriv f x ≤ 0)
+    (h₁ : ∀ x ∈ Set.Ioo b c, 0 ≤ deriv f x) : IsLocalMin f b := by
+    have := localMax_of_deriv_Ioo (f := -f) g₀ g₁
+      (by simp_all) (DifferentiableOn.neg hd₀) (DifferentiableOn.neg hd₁)
+      (fun x hx => deriv.neg (f := f) ▸ Left.nonneg_neg_iff.mpr <|h₀ x hx)
+      (fun x hx => deriv.neg (f := f) ▸ Left.neg_nonpos_iff.mpr <|h₁ x hx)
+    exact (neg_neg f) ▸ IsLocalMax.neg this
+
+/-- If `p` holds to the left of `a` then it holds in an open interval `(l, a)`. -/
+lemma Filter.Eventually.exists_lt_forall_Ioo {α : Type*} [TopologicalSpace α] [LinearOrder α]
+    [OrderTopology α] [NoMinOrder α] {a : α} {p : α → Prop} (h : ∀ᶠ x in 𝓝[<] a, p x) :
+    ∃ l < a, ∀ x ∈ Ioo l a, p x :=
+  mem_nhdsWithin_Iio_iff_exists_Ioo_subset.1 h
+
+/-- If `p` holds to the right of `a` then it holds in an open interval `(a, l)`. -/
+lemma Filter.Eventually.exists_gt_forall_Ioo {α : Type*} [TopologicalSpace α] [LinearOrder α]
+    [OrderTopology α] [NoMaxOrder α] {a : α} {p : α → Prop} (h : ∀ᶠ x in 𝓝[>] a, p x) :
+    ∃ l > a, ∀ x ∈ Ioo a l, p x :=
+  mem_nhdsWithin_Ioi_iff_exists_Ioo_subset.1 h
```

Comment:
just use the existing lemmas and remove these

Reply:
OK done (although it removes an opportunity to use dot notation).

Reply:
In Mathlib we often declare aliases for single implications of bi-implications, especially if it enables dot notation.
That is perfectly fine, and the syntax is `alias ⟨forward_name, backwards_name⟩ := existing_decl` (either name can be `_` to not generate that lemma). Feel free to add that, but it must occur directly after `mem_nhdsWithin_Ioi_iff_exists_Ioo_subset`.

Reply:
Ah okay, I think I'll refrain for now in order to not complicate things.